File: outco/decreaseAndConquer/binarySearch.js
/* 
Binary Search

Leetcode 704: Easy

Successfully passed the test cases on Leetcode on April 1, 2023.

Given an array of integers nums which is sorted in ascending order, 
and an integer target, write a function to search target in nums. 
If target exists, then return its index. Otherwise, return -1.

You must write an algorithm with O(log n) runtime complexity.

Test cases:

Input: nums = [-1,0,3,5,9,12], target = 9
Output: 4
Explanation: 9 exists in nums and its index is 4

Input: nums = [-1,0,3,5,9,12], target = 2
Output: -1
Explanation: 2 does not exist in nums so return -1

Algorithm analysis
Time: O(log(N))
Space: O(1)

*/

const binarySearch = (numsArr, targetInt) => {
    let startPtrInt = 0;
    let endPtrInt = numsArr.length-1;
    while (startPtrInt <= endPtrInt){
        let midPtrInt = parseInt((startPtrInt + endPtrInt)/2)
        if (numsArr[midPtrInt] === targetInt){
            return midPtrInt;
        }
        if (numsArr[midPtrInt] > targetInt){
            endPtrInt = midPtrInt - 1;
        } else {
            startPtrInt = midPtrInt + 1;
        }
    }
    return -1
}

/* TESTS */

const nums1 = [-1,0,3,5,9,12]
const target1 = 9
const test1 = binarySearch(nums1, target1)
console.log(`Test 1: ${test1}`)
// Expected output: Test 2: 4 ✓

const nums2 = [-1,0,3,5,9,12]
const target2 = 2
const test2 = binarySearch(nums2, target2)
console.log(`Test 2: ${test2}`)
// Expected output: Test 2: -1 ✓

const nums3 = [2,5]
const target3 = 5
const test3 = binarySearch(nums3, target3)
console.log(`Test 3: ${test3}`)
// Expected output: Test 3: 1 ✓

const nums4 = [5]
const target4 = 5
const test4 = binarySearch(nums4, target4)
console.log(`Test 4: ${test4}`)
// Expected output: Test 4: 0 ✓

File: c++journey/outco_in_cpp/binarySearch.cpp
 /*  
 *Binary Search

 Given a sorted array and a target value, use binary search to return the index of the target in the input array.
 Return -1 if no such target is found.

 * Parameters
 Input: arr {Array of ints}
 Input: val {int}
 Output: {int}

 *Examples
 [1, 3, 4, 5, 8, 9], 5 --> 3
 [5, 7, 10, 12, 14], 7 --> 1
 [2, 4, 8, 9, 15], 3 --> -1

 * Time Complexity: O(logN)
 * Auxiliary Space Complexity:

 */

#include <iostream>
#include <vector>
#include <string>
using namespace std;

int binarySearch(vector<int> vec, int targetInt) {
    int leftPtrInt = 0;
    int rightPtrInt = vec.size() - 1;
    

    while (leftPtrInt <= rightPtrInt){
        int middlePtrInt = (rightPtrInt + leftPtrInt) / 2;
        if (vec[middlePtrInt] == targetInt){
            return middlePtrInt;
        }
        if (targetInt > vec[middlePtrInt]){
            leftPtrInt = middlePtrInt + 1;
        }
        if (targetInt < vec[middlePtrInt]){
            rightPtrInt = middlePtrInt - 1;
        }
    }
    return -1;
}


/* 
 [1, 3, 4, 5, 8, 9], 5 --> 3
 [5, 7, 10, 12, 14], 7 --> 1
 [2, 4, 8, 9, 15], 3 --> -1

 */
int test(){
    int resultInt1 = binarySearch({1, 3, 4, 5, 8, 9}, 5);
    cout << "Result Integer 1: " << resultInt1 << endl;

    int resultInt2 = binarySearch({5, 7, 10, 12, 14}, 7);
    cout << "\nResult Integer 2: " << resultInt2 << endl;
    

    int resultInt3 = binarySearch({2, 4, 8, 9, 15}, 3);
    cout << "\nResult Integer 3: " << resultInt3 << endl;
    return 0;
}

int main(){
    test();
    return 0;
}

/* Test results 

Result Integer 1: 3

Result Integer 2: 1

Result Integer 3: -1

*/